### PR TITLE
Upgrade to TSLint v4

### PR DESCRIPTION
**Pull request by adidahiya**
#### PR checklist

- [x] New feature, bugfix, or enhancement

#### What changes did you make?

Upgrade TypeScript, TSLint, and tslint-react

#### Is there anything you'd like reviewers to focus on?

~@nchen63 I was able to run the autofixer for a lot of files, but not all of them. It started to hang indefinitely. Can you see if the issue repros for you?~ Update: fixed

### CODE REVIEW

**Comment by nchen63** [comment]
@adidahiya I was able to lint/fix `develop` in 22 seconds

**Comment by adidahiya** [comment]
wow, we OOM'd CircleCI. that is a first 👏

**Comment by adidahiya** [comment]
The problem seems to be a circular symlink of some sort in `package/core`'s node_modules, but builds on master seem to be fine with it..

**Comment by adidahiya** [comment]
ok, now I have new errors that seem to be related to the recent React 15.4 release: https://facebook.github.io/react/blog/2016/11/16/react-v15.4.0.html

- do we have floating dev dependencies on react anywhere? let's not do that.
- the modules `"react/lib/ReactDOM*"` were removed, I wonder if react-day-picker is using those directly

**Comment by adidahiya** [comment]
blocking on #227

**Comment by adidahiya** [comment]
This is mostly linting autofixes; will spare reviewers' time and just get this in. devtested

File: packages/table/src/headers/columnHeader.tsx [hunk 1]
```diff
@@ -77,7 +77,7 @@ export class ColumnHeader extends React.Component<IColumnHeaderProps, {}> {
 
     public render() {
         const { grid, viewportRect, columnIndexStart, columnIndexEnd } = this.props;
-        const cells: React.ReactElement<any>[] = [];
+        const cells: Array<React.ReactElement<any>> = [];
```
Comment (giladgray) [inline packages/table/src/headers/columnHeader.tsx]: @adidahiya what's all this `Array<T>` business? what's wrong with good old-fashioned `T[]`?

Comment (giladgray) [inline packages/table/src/headers/columnHeader.tsx]: but _why_? so i can write `string[]` but not `JSX.Element[]`? I just don't see the logic yet.

Comment (adidahiya) [inline packages/table/src/headers/columnHeader.tsx]: `JSX.Element[]` is fine; check out the rule logic for more specifics. types with generics are non-simple; we deem the readability of `React.ReactElement<any>[]` to be too poor

Comment (giladgray) [inline packages/table/src/headers/columnHeader.tsx]: ok gotcha. that makes more sense: if you're already dealing with generics, use `Array` the same way.

**Pull request merged**